File: src/main/java/com/cuupa/classificator/services/Classificator.java
package com.cuupa.classificator.services;

import com.cuupa.classificator.services.kb.KnowledgeManager;
import com.cuupa.classificator.services.kb.SemanticResult;
import com.cuupa.classificator.services.stripper.PdfAnalyser;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.text.PDFTextStripper;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Classificator {

    private final KnowledgeManager manager;

    private final PdfAnalyser analyser;

    public Classificator(KnowledgeManager manager, PdfAnalyser analyser) {
		this.manager = manager;
        this.analyser = analyser;
    }

    public List<SemanticResult> classify(@Nullable final String text) {
		if(text == null || text.length() == 0) {
			return new ArrayList<>();
		}
		
		return manager.getResults(text);
	}

    @NotNull
    public List<SemanticResult> classify(@Nullable final byte[] content) {
        final List<SemanticResult> results = new ArrayList<>();
        if (content == null || content.length == 0) {
            return results;
        }
        try (PDDocument document = PDDocument.load(new ByteArrayInputStream((content)))) {
            String text = extractText(document).stream().collect(Collectors.joining());
            results.addAll(manager.getResults(text));
            List<SemanticResult> resultFromStructure = analyser.getResults(document);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return results;
    }

    @NotNull
    private List<String> extractText(@NotNull final PDDocument document) {
        final List<String> pages = new ArrayList<>();
        try {
            for (int page = 1; page <= document.getNumberOfPages(); page++) {
                PDFTextStripper stripper = new PDFTextStripper();
                stripper.setStartPage(page);
                stripper.setEndPage(page);
                pages.add(stripper.getText(document));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return pages;
    }
}

File: src/main/java/com/cuupa/classificator/services/Classificator.kt
package com.cuupa.classificator.services

import com.cuupa.classificator.monitor.Monitor
import com.cuupa.classificator.services.kb.KnowledgeManager
import com.cuupa.classificator.services.kb.SemanticResult
import com.cuupa.classificator.services.kb.Topic
import com.cuupa.classificator.services.stripper.PdfAnalyser
import org.apache.juli.logging.LogFactory
import org.apache.pdfbox.pdmodel.PDDocument
import org.apache.pdfbox.text.PDFTextStripper
import java.io.ByteArrayInputStream
import java.io.IOException
import java.time.LocalDateTime
import java.util.*

class Classificator(private val manager: KnowledgeManager, private val analyser: PdfAnalyser,
                    private val monitor: Monitor) {

    fun classify(text: String?): List<SemanticResult> {
        val start = LocalDateTime.now()
        val result = getResultFromInputText(text)
        val done = LocalDateTime.now()
        monitor.writeEvent(manager.getVersion(), text, result, start, done)

        return result
    }

    private fun getResultFromInputText(text: String?): List<SemanticResult> {
        return if (text.isNullOrBlank()) {
            listOf(SemanticResult(Topic.OTHER, mutableListOf()))
        } else {
            manager.getResults(text)
        }
    }

    fun classify(content: ByteArray?): List<SemanticResult> {
        val results: MutableList<SemanticResult> = ArrayList()
        if (content == null || content.isEmpty()) {
            return results
        }
        try {
            PDDocument.load(ByteArrayInputStream(content)).use { document ->
                val text = java.lang.String.join("", extractText(document))
                results.addAll(manager.getResults(text))
                val resultFromStructure = analyser.getResults(document)
            }
        } catch (e: IOException) {
            log.error(e)
        }
        return results
    }

    private fun extractText(document: PDDocument): List<String> {
        val pages: MutableList<String> = ArrayList()
        try {
            for (page in 1..document.numberOfPages) {
                val stripper = PDFTextStripper()
                stripper.startPage = page
                stripper.endPage = page
                pages.add(stripper.getText(document))
            }
        } catch (e: IOException) {
            log.error(e)
        }
        return pages
    }

    companion object {
        private val log = LogFactory.getLog(Classificator::class.java)
    }

}